Take availability.md doc and

davidopp opened the pull request:
- extract the portion related to multi-cluster operation into a new multi-cluster.md doc
- merge the remainder (that was basically high-level troubleshooting advice) into cluster-troubleshooting.md

@erictune @lavalamp

Commit:
Take availability.md doc and
- extract the portion related to multi-cluster operation into a new multi-cluster.md doc
- merge the remainder (that was basically high-level troubleshooting advice) into cluster-troubleshooting.md

lavalamp commented on docs/admin/README.md:
```diff
@@ -45,7 +45,7 @@ Before choosing a particular guide, here are some things to consider:
  - If you are configuring kubernetes on-premises, you will need to consider what [networking
    model](networking.md) fits best.
- - If you are designing for very [high-availability](availability.md), you may want multiple clusters in multiple zones.
+ - If you are designing for very high-availability, you may want [multiple clusters in multiple zones](multi-cluster.md).
```
perhaps just "clusters in multiple zones" is clearer? Since you don't want more than one cluster in a zone, presumably.

davidopp replied:
done

lavalamp commented on docs/admin/cluster-troubleshooting.md:
```diff
@@ -45,7 +45,76 @@ of the relevant log files.  (note that on systemd based systems, you may need to
 ### Worker Nodes
    * /var/log/kubelet.log - Kubelet, responsible for running containers on the node
-   * /var/log/kube-proxy.log - Kube Proxy, responsible for service load balancing
+   * /var/log/kube-proxy.log - Kube Proxy, responsible for service load balancing## A general overview of cluster failure modes
```
missing newlines?

davidopp replied:
done

lavalamp commented on docs/admin/cluster-troubleshooting.md:
```diff
@@ -45,7 +45,76 @@ of the relevant log files.  (note that on systemd based systems, you may need to
+  - network partition within cluster, or between cluster and users.
+  - crashes in Kubernetes software 
+  - data loss or unavailability of persistent storage (e.g. GCE PD or AWS EBS volume).
+  - operator error misconfigures kubernetes software or application software.
```
"operator error, e.g. misconfigured kubernetes software or application software."

Also, make ending with '.' consistent, and maybe capitalize?

davidopp replied:
operator error: done.

removed periods at end of lines (except in a few cases where it seemed necessary) and sort of made the capitalization more consistent

lavalamp commented on docs/admin/cluster-troubleshooting.md:
```diff
@@ -45,7 +45,76 @@ of the relevant log files.  (note that on systemd based systems, you may need to
+    - Results
+      - unable to stop, update, or start new pods, services, replication controller
+      - existing pods and services should continue to work normally, unless they depend on the Kubernetes API
+  - Apiserver backing storage lost
```
insert "(i.e. etcd's data directory)"

davidopp replied:
done

lavalamp commented on docs/admin/cluster-troubleshooting.md:
```diff
@@ -45,7 +45,76 @@ of the relevant log files.  (note that on systemd based systems, you may need to
+  - Supporting services (node controller, replication controller manager, scheduler, etc) VM shutdown or crashes
+    - currently those are colocated with the apiserver, and their unavailability has similar consequences as apiserver
+    - in future, these will be replicated as well and may not be co-located
+    - they do not have own persistent state
```
s/own/their own/

davidopp replied:
done

lavalamp commented on docs/admin/cluster-troubleshooting.md:
```diff
@@ -45,7 +45,76 @@ of the relevant log files.  (note that on systemd based systems, you may need to
+    - currently those are colocated with the apiserver, and their unavailability has similar consequences as apiserver
+    - in future, these will be replicated as well and may not be co-located
+    - they do not have own persistent state
+  - Node (thing that runs kubelet and kube-proxy and pods) shutdown
```
"individual node is shutdown"

side note, I feel like this use of "shutdown" is weird, "shutdown" is a verb to me; I would just say "down", "off", "gone", or "terminated" when a noun is required. But maybe this is just me.

davidopp replied:
changed to "individual node is shut down"

davidopp replied:
I mean "individual node is shut down"

lavalamp commented on docs/admin/cluster-troubleshooting.md:
```diff
@@ -45,7 +45,76 @@ of the relevant log files.  (note that on systemd based systems, you may need to
+    - they do not have own persistent state
+  - Node (thing that runs kubelet and kube-proxy and pods) shutdown
+    - Results
+      - pods on that Node stop running
```
Maybe insert:
```
  - Network partition
    - Results
      - Partition A thinks the nodes in the partition B are down; Partition B thinks the apiserver is down. (Assuming the master VM ends up in partition A.)

davidopp replied:
done

lavalamp commented on docs/admin/cluster-troubleshooting.md:
```diff
@@ -45,7 +45,76 @@ of the relevant log files.  (note that on systemd based systems, you may need to
+  - Node (thing that runs kubelet and kube-proxy and pods) shutdown
+    - Results
+      - pods on that Node stop running
+  - Kubelet software fault
```
Kubelet or docker?

davidopp replied:
These are Kubelet, right? Or am I missing something?

lavalamp commented on docs/admin/cluster-troubleshooting.md:
```diff
@@ -45,7 +45,76 @@ of the relevant log files.  (note that on systemd based systems, you may need to
+      - kubelet might delete the pods or not
+      - node marked unhealthy
+      - replication controllers start new pods elsewhere
+  - Cluster operator error
```
Isn't this section really super obvious?

davidopp replied:
Yes, but it might remind someone to consider whether *they* are the problem. :)

erictune replied:
I originally put this section in because I wanted people to think about how operator error can make a single multi-zone cluster significantly less available than multiple clusters.

lavalamp replied:
Ah, that makes sense.... a bit less so now that it's moved out of that doc :)

lavalamp commented on docs/admin/cluster-troubleshooting.md:
```diff
@@ -45,7 +45,76 @@ of the relevant log files.  (note that on systemd based systems, you may need to
+- Action use IaaS providers reliable storage (e.g GCE PD or AWS EBS volume) for VMs with apiserver+etcd.
+  - Mitigates: Apiserver backing storage lost
+
+- Action: Use Replicated APIserver feature (when complete: feature is planned but not implemented)
```
s/Replicated/replicated/

lavalamp replied:
Also link to the availability doc?

davidopp replied:
done and done.

davidopp commented:
Thanks for the review; PTAL.

BTW the "general overview of cluster failure modes" section pre-existed (and I didn't write it); I just moved it from one doc to another. But you had good feedback and I implemented your suggestions.

lavalamp commented:
LGTM with squash :)

davidopp commented:
Squashed

davidopp commented:
@k8s-bot test this please

erictune commented:
@davidopp I didn't write the failure modes section with troubleshooting in mind.  I wrote it to help people to think about all the factors that affect availability, prior to talking about multi-zone clusters vs multi-cluster.    

So, I haven't looked closely at the the changes, but I think a copy-and-edit would have been better than a move.

davidopp commented:
@erictune My apologies. I did not see the connection the first time I read the doc. The list of problems and mitigations you gave looks quite general, which is why I moved it into the troubleshooting doc. AFAICT only two of the things on the list are related to HA (Use replicated APIserver, and Multiple independent clusters). I think it's an excellent list, and that it will be more discoverable and useful if it's part of the troubleshooting doc. So even though I misunderstood your original intention, I still think the new organization is better. But if you disagree, feel free to send a PR moving it back (but if you do so, please add more connective tissue between the list of problems and mitigations, and the description of multi-cluster configurations).
